### PR TITLE
Generate Contract Types (with Code)

### PR DESCRIPTION
**Pull request by ricklove**
This uses the type generator and produces code output. 

It modifies `michelson-to-ts` to generate the json version of the contract then to call generate-types which generates the code and types for the contract.

Differences:

- code is a string instead of being wrapped in JSON.parse 
    - JSON.parse causes the string to be parsed at import instead of at runtime
    - TezosToolkit.originate will parse the string (in fact it will parse the original michelson code, so this could just be a copy of the .tz file)
    - If the JSON object is needed, it's simple to call JSON.parse(code)
- filenames have changed
- uses named exports which is friendly for ide discovery and auto import 
    - just type the name of the contract in Pascal case and it will be auto imported by vscode, etc. 
    - this makes tree shaking unnecessary which is more friendly for simple build setups
- contract types and code are in separate files
    - code is only needed for origination and often not at runtime, so this large file can be optionally imported only when needed

### CODE REVIEW

File: packages/minter-contracts/src/michelson-to-ts.ts [hunk 1]
```diff
@@ -66,43 +65,67 @@ const compile = (contracts: string[]) =>
...
-    const compiledFiles = await compile(contracts);
+
+    // Compile json
+    await compileToJsonFiles(contracts);
+
+    // Generate Types
+    await generateTypes();
+
+    const outputFiles = await fs.promises.readdir(outPath);
+    const codeFiles = outputFiles
```
Comment (zachrbrown) [inline packages/minter-contracts/src/michelson-to-ts.ts:124]: Do we need to poll the directory again here? I believe the `contracts` variable would suffice here (and would save a `filter` call).

File: packages/minter-contracts/README.md [hunk 2]
```diff
@@ -116,6 +116,15 @@ E.g.,
 yarn michelson-to-ts -c tezos-client
 ```
 
+### Generate Contract Types (this is run as part of michelson-to-ts)
```
Comment (zachrbrown) [inline packages/minter-contracts/README.md:119]: To follow suit with other commands, the command name should be the header.
```markdown
### `yarn generate-types`
```

And the table of contents should be updated to include this as well.

**Review by zachrbrown — CHANGES_REQUESTED** [review]
Overall looks good! Few comments.

File: packages/minter-contracts/src/michelson-to-ts.ts [hunk 3]
```diff
@@ -66,43 +65,67 @@ const compile = (contracts: string[]) =>
...
+    .map(fileName => `  ${fileNameToContractType(fileName)},\n  ${fileNameToCodeObject(fileName)},`)
     .join('\n');
 
-  return `${imports}\n\nexport {\n${exports}\n};\n`;
+  return `${codeImports}\n\nexport {\n${exports}\n};\n`;
 };
 
 (async () => {
   try {
+    fs.promises.mkdir(outPath);
```
[the remark this answers is not in the document]
Comment (ricklove) [reply packages/minter-contracts/src/michelson-to-ts.ts:106]: Ah, never noticed this before: `Calling fsPromises.mkdir() when path is a directory that exists results in a rejection only when recursive is false.`

File: packages/minter-contracts/bin-ts/english_auction_fa2.code.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+
+export const EnglishAuctionFa2Code: { __type: 'EnglishAuctionFa2Code', code: string } = {
```
[the remark this answers is not in the document]
Comment (ricklove) [reply packages/minter-contracts/bin-ts/english_auction_fa2.code.ts:2]: Fair enough

File: packages/minter-contracts/src/michelson-to-ts.ts (hunk 1 above)
Comment (ricklove) [reply packages/minter-contracts/src/michelson-to-ts.ts:124]: This is a different directory than the original

**Comment by ricklove** [comment]
This is replaced by: #93

**Comment by ricklove** [comment]
Replace by #93

**Pull request closed without merging**